tests/logic/test_submit_logic.py

```diff
@@ -0,0 +1,176 @@
+
+    test_start_pipeline_run_request_body = StartPipelineRunRequestBody(
+        description=test_description, jobControl=JobControl(id=test_job_id)
+    )
+    mock_job_report = mock(
+        {"id": test_job_id, "status_code": 202}
+    )  # successful (running) status code
+    mock_async_pipeline_run_response = mock({"job_report": mock_job_report})
+    when(mock_pipeline_runs_api).start_pipeline_run(
+        test_pipeline_name, test_start_pipeline_run_request_body
+    ).thenReturn(mock_async_pipeline_run_response)
+
+    # Act
+    result = submit_logic.start_pipeline_run(
+        test_pipeline_name, test_job_id, test_description
+    )
+
+    # Assert
+    assert result == test_job_id
+    verify(mock_pipeline_runs_api).start_pipeline_run(
+        test_pipeline_name, test_start_pipeline_run_request_body
+    )
+
+
+def test_start_pipeline_run_error_response(mock_pipeline_runs_api):
+    # Arrange
+    test_job_id = str(uuid.uuid4())
+    test_pipeline_name = "foobar"
+    test_description = "user-provided description"
+
+    test_start_pipeline_run_request_body = StartPipelineRunRequestBody(
+        description=test_description, jobControl=JobControl(id=test_job_id)
+    )
+    mock_job_report = mock(
+        {"id": test_job_id, "status_code": 500}
+    )  # internal error status code
+    mock_error_report = mock({"message": "some error message"})
+    mock_async_pipeline_run_response = mock(
+        {"job_report": mock_job_report, "error_report": mock_error_report}
+    )
+    when(mock_pipeline_runs_api).start_pipeline_run(
+        test_pipeline_name, test_start_pipeline_run_request_body
+    ).thenReturn(mock_async_pipeline_run_response)
+
+    # Act
+    response = submit_logic.start_pipeline_run(
+        test_pipeline_name, test_job_id, test_description
+    )
+
+    # Assert
+    assert response == test_job_id
+    verify(mock_pipeline_runs_api).start_pipeline_run(
+        test_pipeline_name, test_start_pipeline_run_request_body
+    )
+
+
+def test_prepare_upload_start_pipeline_run():
+    # Arrange
+    test_pipeline_name = "foobar"
+    test_pipeline_version = 0
+    test_input_name = "input_name"
+    test_input_value = "value"
+    test_inputs = {test_input_name: test_input_value}
+    test_description = "user-provided description"
+
+    test_job_id = uuid.uuid4()
+    test_job_id_str = str(test_job_id)
+    when(submit_logic.uuid).uuid4().thenReturn(test_job_id)
+
+    test_signed_url = "signed_url"
+    test_upload_url_dict = {test_input_name: test_signed_url}
+    when(submit_logic).prepare_pipeline_run(
+        test_pipeline_name, test_job_id_str, test_pipeline_version, test_inputs
+    ).thenReturn(test_upload_url_dict)
+
+    when(submit_logic).upload_file_with_signed_url(
+        test_input_value, test_signed_url
+    )  # do nothing
+
+    when(submit_logic).start_pipeline_run(
+        test_pipeline_name, test_job_id_str, test_description
+    ).thenReturn(test_job_id)
+
+    # Act
```

Comment:
what are these comments of `Act` and `Arrange` and stuff?

Reply:
they were how some of the autogenerated tests organized the testing code. Arrange is the variables and mocks, Act is actually calling the thing you're testing, and Assert is the assertions. happy to remove them if they're dumb